### PR TITLE
feat!: Add acceptable code for request and exec

### LINKED ISSUE #188: [bug]: a `request` command that returns a non-200 exit code still passes
### What version of the Kurtosis CLI were you running when you encountered this bug?

0.68.8

### (Optional) What is your metrics user ID?

0497e811755b8f775ef0d99cd22b1ef221708b5a95119857a310cbc4f2eabcae

### What is the severity of this bug?

Painful, this is causing significant friction in my workflow.

### Description & steps to reproduce

Run the following package:
[package.zip](https://github.com/kurtosis-tech/kurtosis/files/10972220/package.zip)

### Relevant log output

```true
> request recipe=PostHttpRequestRecipe(port_id="http", endpoint="/actor", body="{\"first_name\":\"Kevin\",\"last_name\":\"Bacon\"}", content_type="application/json", extract="")
Request had response code '503' and body "{\"code\":\"PGRST002\",\"details\":null,\"hint\":null,\"message\":\"Could not query the database for the schema cache. Retrying.\"}"
```

### Desired behavior

A non-200 should, by default, be a failed Starlark instruction

### LINKED ISSUE #201: [bug]: An `exec` that returns a non-0 exit code still passes
### What version of the Kurtosis CLI were you running when you encountered this bug?

0.68.8

### (Optional) What is your metrics user ID?

_No response_

### What is the severity of this bug?

Painful, this is causing significant friction in my workflow.

### Description & steps to reproduce

Run an `exec` that returns a non-0 exit code

### Relevant log output

```true
> exec recipe=ExecRecipe(service_name="postgres", command=["pg_restore", "-U", "app_user", "-d", "app_db", "--no-owner", "--role=app_user", "/seed-data/dvdrental.tar"])
Command returned with exit code '1' and the following output:
--------------------
pg_restore: error: could not open input file "/seed-data/dvdrental.tar": No such file or directory

--------------------
```

### Desired behavior

The step should fail at execution

### PR DESCRIPTION
**Pull request by victorcolombo**
## Is this change user facing?

* [X] Yes
* [ ] No

## References (if applicable):

Closes #201
Closes #188

### CODE REVIEW

File: core/server/api_container/server/startosis_engine/kurtosis_instruction/exec/exec.go [hunk 1]
```diff
@@ -118,6 +145,16 @@ func (builtin *ExecCapabilities) Execute(ctx context.Context, _ *builtin_argumen
...
 	}
+	isAcceptableCode := false
+	for _, acceptableCode := range builtin.acceptableCodes {
+		if result["code"] == starlark.MakeInt(acceptableCode) {
+			isAcceptableCode = true
+			break
+		}
+	}
+	if !isAcceptableCode {
+		return "", stacktrace.NewError("Request returned status code '%v' that is not part of the acceptable status codes '%v'", result["code"], builtin.acceptableCodes)
```
Comment (h4ck3rk3y) [inline core/server/api_container/server/startosis_engine/kurtosis_instruction/exec/exec.go]: nit: s/status/exit/ 

as this is command line and not requests

Comment (h4ck3rk3y) [reply core/server/api_container/server/startosis_engine/kurtosis_instruction/exec/exec.go]: s/Request/exec

File: core/server/api_container/server/startosis_engine/kurtosis_types/arg_shared_helpers.go [hunk 2]
```diff
@@ -66,3 +67,20 @@ func SafeCastToString(expectedValueString starlark.Value, argNameForLogging stri
...
+		value := starlarkList.Index(i)
+		starlarkCastedValue, ok := value.(starlark.Int)
+		if !ok {
+			return nil, stacktrace.NewError("An error occurred when casting element '%v' from slice '%v' to integer", value, starlarkList)
+		}
+		castedValue, ok := starlarkCastedValue.Int64()
+		if !ok {
+			return nil, stacktrace.NewError("An error occurred when casting element '%v' to Go integer", castedValue)
+		}
+		slice = append(slice, int(castedValue))
```
Comment (h4ck3rk3y) [inline core/server/api_container/server/startosis_engine/kurtosis_types/arg_shared_helpers.go]: is there an overflow that could happen here?

File: internal_testsuites/golang/testsuite/starlark_exec_test/starlark_exec_test.go [hunk 3]
```diff
@@ -40,7 +40,7 @@ def run(plan, args):
 		service_name = "test",
 		command = %v,
 	)
-	exec_result = plan.exec(exec_recipe)
+	exec_result = plan.exec(exec_recipe, acceptable_codes=[0, 1])
```
Comment (h4ck3rk3y) [inline internal_testsuites/golang/testsuite/starlark_exec_test/starlark_exec_test.go]: is this non deterministic?

**Review by h4ck3rk3y — APPROVED** [review]
LGTM functionally, have some comments

1. This should be a breaking change if Waku/Eth use something like
```
exec_response = plan.exec()
# this should fail
plan.assert(exec_response["code"], "==", "-1")
```

2. This requires the user to know exactly what the failing exit code would be which might be too much. Its tough to know how a particular thing would fail on a given image, I can't do something like

```
exec_response = plan.exec()
# anything but success is okay
plan.assert(exec_response["code"], "!=", "0")
```

File: internal_testsuites/golang/testsuite/starlark_exec_test/starlark_exec_test.go (hunk 3 above)
Comment (victorcolombo) [reply internal_testsuites/golang/testsuite/starlark_exec_test/starlark_exec_test.go]: We run this test over multiple scripts (it's parametrized) and do the assert later. It's also a nice way to test the feature

File: core/server/api_container/server/startosis_engine/kurtosis_instruction/exec/exec.go [hunk 4]
```diff
@@ -17,11 +18,21 @@ import (
...
+
 const (
 	ExecBuiltinName = "exec"
 
-	RecipeArgName      = "recipe"
-	ServiceNameArgName = "service_name"
+	RecipeArgName          = "recipe"
+	ServiceNameArgName     = "service_name"
+	AcceptableCodesArgName = "acceptable_codes"
+	SkipCodeCheckArgName   = "skip_code_check"
```
Comment (h4ck3rk3y) [inline core/server/api_container/server/startosis_engine/kurtosis_instruction/exec/exec.go:31]: code/codes ?

File: core/server/api_container/server/startosis_engine/kurtosis_instruction/exec/exec.go [hunk 5]
```diff
@@ -92,6 +119,27 @@ func (builtin *ExecCapabilities) Interpret(arguments *builtin_argument.ArgumentV
...
+		if err != nil {
+			return nil, startosis_errors.WrapWithInterpretationError(err, "Unable to parse '%v' argument", acceptableCodes)
+		}
+	}
+
+	skipCodeCheck := defaultSkipCodeCheck
+	if arguments.IsSet(SkipCodeCheckArgName) {
+		skipCodeCheckArgumentValue, err := builtin_argument.ExtractArgumentValue[starlark.Bool](arguments, SkipCodeCheckArgName)
+		if err != nil {
+			return nil, startosis_errors.WrapWithInterpretationError(err, "Unable to extract value for '%s' argument", ServiceNameArgName)
```
Comment (h4ck3rk3y) [inline core/server/api_container/server/startosis_engine/kurtosis_instruction/exec/exec.go]: is this the right arg name?

File: core/server/api_container/server/startosis_engine/kurtosis_instruction/exec/exec.go (hunk 4 above)
Comment (h4ck3rk3y) [reply core/server/api_container/server/startosis_engine/kurtosis_instruction/exec/exec.go:31]: lets add docs for these

File: core/server/api_container/server/startosis_engine/kurtosis_instruction/request/request.go [hunk 6]
```diff
@@ -8,20 +8,41 @@ import (
...
 
 const (
 	RequestBuiltinName = "request"
 
-	RecipeArgName      = "recipe"
-	ServiceNameArgName = "service_name"
+	RecipeArgName          = "recipe"
+	ServiceNameArgName     = "service_name"
+	AcceptableCodesArgName = "acceptable_codes"
+	SkipCodeCheckArgName   = "skip_code_check"
```
Comment (h4ck3rk3y) [inline core/server/api_container/server/startosis_engine/kurtosis_instruction/request/request.go:45]: code/codes?

File: core/server/api_container/server/startosis_engine/kurtosis_instruction/request/request.go [hunk 7]
```diff
@@ -93,6 +142,15 @@ func (builtin *RequestCapabilities) Interpret(arguments *builtin_argument.Argume
 		return nil, startosis_errors.NewInterpretationError("Service name is not set, either as a request instruction's argument or as a recipe field. You can fix it passing the 'service_name' argument in the 'request' call")
 	}
 
+	skipCodeCheck := defaultSkipCodeCheck
+	if arguments.IsSet(SkipCodeCheckArgName) {
+		skipCodeCheckArgumentValue, err := builtin_argument.ExtractArgumentValue[starlark.Bool](arguments, SkipCodeCheckArgName)
+		if err != nil {
+			return nil, startosis_errors.WrapWithInterpretationError(err, "Unable to extract value for '%s' argument", ServiceNameArgName)
```
Comment (h4ck3rk3y) [inline core/server/api_container/server/startosis_engine/kurtosis_instruction/request/request.go]: SkipCodeCheckArgName

File: core/server/api_container/server/startosis_engine/kurtosis_instruction/request/request.go [hunk 8]
```diff
@@ -118,7 +178,21 @@ func (builtin *RequestCapabilities) Execute(ctx context.Context, _ *builtin_argu
...
 	}
+	if !builtin.skipCodeCheck && !builtin.isAcceptableCode(result) {
+		return "", stacktrace.NewError("Request returned status code '%v' that is not part of the acceptable status codes '%v'", result["code"], builtin.acceptableCodes)
+	}
 	builtin.runtimeValueStore.SetValue(builtin.resultUuid, result)
 	instructionResult := builtin.httpRequestRecipe.ResultMapToString(result)
 	return instructionResult, err
 }
+
+func (builtin *RequestCapabilities) isAcceptableCode(recipeResult map[string]starlark.Comparable) bool {
```
Comment (h4ck3rk3y) [inline core/server/api_container/server/startosis_engine/kurtosis_instruction/request/request.go:189]: could be put in a helper as its same but maybe its overdrying

Comment (h4ck3rk3y) [reply core/server/api_container/server/startosis_engine/kurtosis_instruction/request/request.go:189]: why doesn't this language have a `in` primitive

File: internal_testsuites/golang/testsuite/starlark_exec_test/starlark_exec_test.go [hunk 9]
```diff
@@ -40,7 +40,7 @@ def run(plan, args):
 		service_name = "test",
 		command = %v,
 	)
-	exec_result = plan.exec(exec_recipe)
+	exec_result = plan.exec(exec_recipe, acceptable_codes=[0], skip_code_check=True)
```
Comment (h4ck3rk3y) [inline internal_testsuites/golang/testsuite/starlark_exec_test/starlark_exec_test.go:43]: do we have to pass "acceptable_codes"??

File: docs/docs/reference/starlark-instructions.md [hunk 10]
```diff
@@ -175,10 +175,18 @@ exec_recipe = ExecRecipe(
...
+    # The recipe that will determine the exec to be performed.
     # Valid values are of the following types: (ExecRecipe)
     # MANDATORY
     recipe = exec_recipe,
+    
+    # If the recipe returns a code that does not belong on this list, this instruction will fail.
+    # OPTIONAL (Defaults to [0])
+    acceptable_codes = [0, 0], # Here both 0 and 1 are valid codes that we want to accept and not fail the instruction
+    
+    # If False, instruction will never fail based on code. You can chain this call with assert to check codes after request is done.
```
Comment (h4ck3rk3y) [inline docs/docs/reference/starlark-instructions.md]: lets be explicit that setting this to true makes acceptable_codes useless

**Review by h4ck3rk3y — APPROVED** [review]

File: docs/docs/reference/starlark-instructions.md (hunk 10 above)
Comment (victorcolombo) [reply docs/docs/reference/starlark-instructions.md]: Done

File: core/server/api_container/server/startosis_engine/kurtosis_instruction/exec/exec.go (hunk 4 above)
Comment (victorcolombo) [reply core/server/api_container/server/startosis_engine/kurtosis_instruction/exec/exec.go:31]: Added docs for both `exec` and `request`

File: core/server/api_container/server/startosis_engine/kurtosis_instruction/request/request.go (hunk 6 above)
Comment (victorcolombo) [reply core/server/api_container/server/startosis_engine/kurtosis_instruction/request/request.go:45]: I believe "Code Check" sounds more reasonable than Codes Check

File: core/server/api_container/server/startosis_engine/kurtosis_types/arg_shared_helpers.go (hunk 2 above)
Comment (victorcolombo) [reply core/server/api_container/server/startosis_engine/kurtosis_types/arg_shared_helpers.go]: Switched to `int64` for safety

File: core/server/api_container/server/startosis_engine/kurtosis_instruction/request/request.go (hunk 8 above)
Comment (victorcolombo) [reply core/server/api_container/server/startosis_engine/kurtosis_instruction/request/request.go:189]: in this case I think it's reasonable to leave this here, given how trivial it is and there is no good way to place it

**Pull request merged**